### PR TITLE
api: generalize version information to any platform component

### PR DESCRIPTION
**Pull request by tiborvass**
This change adds a Platform struct with a Name field and a general
Components field to the Version API type. This will allow API
consumers to show version information for the whole platform and
it will allow API providers to set the versions for the various
components of the platform.

All changes here are backwards compatible.

Signed-off-by: Tibor Vass <tibor@docker.com>

### CODE REVIEW

**Comment by friism** [comment]
I agree with @mavenugo - a switch on sub-commands where kube support works is easier to reason about than env vars. Esp. since not all the orchestration-related commands (`docker stack`, `docker service`, `docker node`) will understand the env var out of the gate. It's confusing for a user if they set the env var and then only some orchestrator commands interact with kube. A per-command switch makes that clearer.

I think we have to content with this dual-orchestrator model for a while.

File: daemon/info.go [hunk 1]
```diff
@@ -155,14 +155,16 @@ func (daemon *Daemon) SystemInfo() (*types.Info, error) {
...
 	v := types.Version{
-		Version:       dockerversion.Version,
-		GitCommit:     dockerversion.GitCommit,
-		MinAPIVersion: api.MinVersion,
-		GoVersion:     runtime.Version(),
-		Os:            runtime.GOOS,
-		Arch:          runtime.GOARCH,
-		BuildTime:     dockerversion.BuildTime,
-		Experimental:  daemon.configStore.Experimental,
+		types.EngineVersion{
```
Comment (cpuguy83) [inline daemon/info.go]: Needs an field identifier `EngineVersion: types.EngineVersion{}`

**Review by cpuguy83 — CHANGES_REQUESTED** [review]

**Review by dnephin — COMMENTED** [review]
Design LGTM

**Comment by friism** [comment]
> All changes here are backwards compatible.

I think this change is good but just want to note that some API clients will fail on "unknown" JSON properties (like the one introduced here). For evidence, Google for `UnrecognizedPropertyException`. A related class of problems is that intermediary clients or API proxies may not error, but may scrub unknown properties from responses that they de-serialize and re-serialize as part of passing responses back and forth.

**Comment by tiborvass** [comment]
@friism thanks, that is fine I think because we will remove the old fields from the documentation for API 1.35

**Comment by cpuguy83** [comment]
@friism This is a problem for basically every API change we make because we don't send old versions of types to older clients, so it can always have extra fields in it... however we do ensure that older fields are available and populated as expected.

**Comment by cpuguy83** [comment]
And wow really sorry for the late response... I guess I got served a cached version of this page and din't get tibor's response until just now...

File: api/swagger.yaml [hunk 2]
```diff
@@ -6764,6 +6764,26 @@ paths:
...
+                required: [Name]
+                properties:
+                  Name:
+                    type: "string"
+              Components:
+                type: "array"
+                items:
+                  type: "object"
+                  x-go-name: ComponentVersion
+                  required: [Version]
```
Comment (thaJeztah) [inline api/swagger.yaml]: Name is required here as well, correct?

File: api/swagger.yaml [hunk 3]
```diff
@@ -6764,6 +6764,26 @@ paths:
...
+                  properties:
+                    Version:
```
Comment (thaJeztah) [inline api/swagger.yaml]: Name property is missing

**Review by thaJeztah — CHANGES_REQUESTED** [review]

File: daemon/info.go [hunk 4]
```diff
@@ -153,25 +153,47 @@ func (daemon *Daemon) SystemInfo() (*types.Info, error) {
...
+	}
+
 	v := types.Version{
+		Components: []types.ComponentVersion{
+			{
+				Name:    "Engine",
+				Version: dockerversion.Version,
+				Details: map[string]string{
+					"GitCommit":     dockerversion.GitCommit,
+					"ApiVersion":    APIVersion,
```
Comment (thaJeztah) [inline daemon/info.go]: Still confused why we can't access `api.DefaultVersion` here directly

Comment (tiborvass) [reply daemon/info.go]: I assumed a circular dependency, but it's true i didnt try.

**Review by cpuguy83 — APPROVED** [review]
LGTM

**Review by thaJeztah — APPROVED** [review]
LGTM, thanks!!

**Comment by thaJeztah** [comment]
Janky failure is not related;

```
23:37:23 ----------------------------------------------------------------------
23:37:23 FAIL: check_test.go:366: DockerSwarmSuite.TearDownTest
23:37:23 
23:37:23 check_test.go:371:
23:37:23     d.Stop(c)
23:37:23 daemon/daemon.go:395:
23:37:23     t.Fatalf("Error while stopping the daemon %s : %v", d.id, err)
23:37:23 ... Error: Error while stopping the daemon d35b5ad05f17b : exit status 130
23:37:23 
23:37:23 
23:37:23 ----------------------------------------------------------------------
23:37:23 PANIC: docker_api_swarm_service_test.go:201: DockerSwarmSuite.TestAPISwarmServicesUpdateStartFirst
23:37:23 
23:37:23 [d35b5ad05f17b] waiting for daemon to start
23:37:23 [d35b5ad05f17b] daemon started
23:37:23 
23:37:23 [d35b5ad05f17b] daemon started
23:37:23 Attempt #2: daemon is still running with pid 8909
23:37:23 Attempt #3: daemon is still running with pid 8909
23:37:23 Attempt #4: daemon is still running with pid 8909
23:37:23 [d35b5ad05f17b] exiting daemon
23:37:23 ... Panic: Fixture has panicked (see related PANIC)
23:37:23 
23:37:23 ----------------------------------------------------------------------
```

**Comment by thaJeztah** [comment]
Merging, as the failure is not related

**Comment by stevvooe** [comment]
How will this be disambiguated from the concept of a platform, which includes the definition of an architecture and operating system?

**Comment by tianon** [comment]
Agree with Steve, calling this "platform" is an unfortunate choice given what it's defined to mean elsewhere 😥

**Pull request merged**